homeassistant/components/airvisual/__init__.py

```diff
@@ -1 +1,220 @@
 """The airvisual component."""
+import asyncio
+from datetime import timedelta
+import logging
+
+from pyairvisual import Client
+from pyairvisual.errors import AirVisualError, InvalidKeyError
+import voluptuous as vol
+
+from homeassistant.config_entries import SOURCE_IMPORT
+from homeassistant.const import (
+    CONF_API_KEY,
+    CONF_LATITUDE,
+    CONF_LONGITUDE,
+    CONF_SCAN_INTERVAL,
+    CONF_SHOW_ON_MAP,
+    CONF_STATE,
+)
+from homeassistant.core import callback
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import aiohttp_client, config_validation as cv
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.event import async_track_time_interval
+
+from .config_flow import configured_instances
+from .const import (
+    CONF_CITY,
+    CONF_COUNTRY,
+    CONF_GEOGRAPHIES,
+    DATA_CLIENT,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    TOPIC_OPTIONS_UPDATE,
+    TOPIC_UPDATE,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_LISTENER = "listener"
+
+CONF_NODE_ID = "node_id"
+
+GEOGRAPHY_SCHEMA = vol.All(
+    vol.Schema(
+        {
+            vol.Inclusive(CONF_CITY, "city"): cv.string,
+            vol.Inclusive(CONF_STATE, "city"): cv.string,
+            vol.Inclusive(CONF_COUNTRY, "city"): cv.string,
+            vol.Inclusive(CONF_LATITUDE, "coords"): cv.latitude,
+            vol.Inclusive(CONF_LONGITUDE, "coords"): cv.longitude,
+            vol.Optional(CONF_SHOW_ON_MAP, default=True): cv.boolean,
+        },
+        cv.has_at_least_one_key(CONF_CITY, CONF_LATITUDE),
+    )
+)
+
+CLOUD_API_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_API_KEY): cv.string,
+        vol.Optional(CONF_GEOGRAPHIES, default=[]): vol.All(
+            cv.ensure_list, [GEOGRAPHY_SCHEMA]
+        ),
+        vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL): vol.All(
+            cv.time_period, lambda value: value.total_seconds()
+        ),
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: CLOUD_API_SCHEMA}, extra=vol.ALLOW_EXTRA,)
+
+
+@callback
+def async_get_geography_id(geography_dict):
+    """Generate a unique ID from a geography_dict dict."""
+    if CONF_CITY in geography_dict:
+        return ",".join(
+            (
+                geography_dict[CONF_CITY],
+                geography_dict[CONF_STATE],
+                geography_dict[CONF_COUNTRY],
+            )
+        )
+    return ",".join(
+        (str(geography_dict[CONF_LATITUDE]), str(geography_dict[CONF_LONGITUDE]))
+    )
+
+
+async def async_setup(hass, config):
+    """Set up the AirVisual component."""
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][DATA_CLIENT] = {}
+    hass.data[DOMAIN][DATA_LISTENER] = {}
+
+    if DOMAIN not in config:
+        return True
+
+    conf = config[DOMAIN]
+
+    if conf[CONF_API_KEY] in configured_instances(hass):
```

Comment:
Why support configuration.yaml At all when you introduce a breaking change on the configuration?

Reply:
I still want to give control to people who prefer managing their configuration in YAML.